gradio/documentation.py

```diff
@@ -0,0 +1,146 @@
+import inspect
+
+classes_to_document = {}
+gl = {"mode": None}
+
+
+def document_mode(m):
+    gl["mode"] = m
+    if m not in classes_to_document:
+        classes_to_document[m] = []
+
+
+def document(*fns):
+    def inner_doc(cls):
+        classes_to_document[gl["mode"]].append((cls, fns))
+        return cls
+
+    return inner_doc
+
+
+def document_fn(fn):
+    doc_str = inspect.getdoc(fn)
+    doc_lines = doc_str.split("\n")
+    signature = inspect.signature(fn)
+    description, parameters, returns, examples = [], {}, [], []
+    mode = "description"
+    for line in doc_lines:
+        if line == "Parameters:":
+            mode = "parameter"
+        elif line == "Example:":
+            mode = "example"
+        elif line == "Returns:":
+            mode = "return"
+        else:
+            if mode == "description":
+                description.append(line if line != "" else "<br>")
+                continue
+            if line.startswith("    "):
+                line = line[4:]
+            if mode == "parameter":
+                colon_index = line.index(": ")
+                parameter = line[:colon_index]
+                parameter_doc = line[colon_index + 2 :]
+                parameters[parameter] = parameter_doc
+            elif mode == "return":
+                returns.append(line)
+            elif mode == "example":
+                examples.append(line)
+    description_doc = " ".join(description)
+    parameter_docs = []
+    for param_name, param in signature.parameters.items():
```

Comment:
We probably want to try/catch some of the internal logic in case one of the parameters is misspelled in the docstring or something.

Reply:
Disagree, generate_documentation should error out if there's an issue with documentation. Will add a test for generate_documentation to our unit tests so bad documentation never gets merged.